Performance.md

```diff
@@ -0,0 +1,290 @@
+              |  Shared  |
+              ^-----^----^
+             /      |     \
+            /       |      \
+------------  ------------  ------------
+|          |  |  Shared  |  |          |
+|  Client  |  |  Tests   |  |  Server  |
+-----^------  ------------  ------^-----
+     |                            |
+     |                            |
+------------                ------------
+|  Client  |                |  Server  |
+|  Tests   |                |  Tests   |
+------------                ------------
+```
+
+You can read up more about project references here. <!-- TODO -->
+
+# Configuring `tsconfig.json`
+
+Within a `tsconfig.json`, there are two ways to specify files in a project:
+
+* the `files` list
+* the `include` and `exclude` lists
+
+The primary difference between the two is that `files` expects a list of file paths to source files, and `include/`exclude` use globbing patterns to match against files.
+
+While specifying `files` will allow TypeScript to quickly load up files up directly, it can be cumbersome if you have many files in your project without just a few top-level entry-points.
+Additionally, it's easy to forget to add new files to your `tsconfig.json`, which means that you might end up with strange editor behavior where those new files are incorrectly analyzed.
+All this can be cumbersome.
+
+`include`/`exclude` help avoid needing to specify these files, but at a cost: files must be discovered by walking through included directories.
+When running through a *lot* of folders, this can slow compilations down.
+Additionally, sometimes a compilation will include lots of unnecessary `.d.ts` files and test files, which can increase compilation time and memory overhead.
+Finally, while `exclude` has some reasonable defaults, certain configurations like mono-repos mean that a "heavy" folders like `node_modules` can still end up being included.
+
+For best practices, we recommend the following:
+
+* Specify only source folders in your project.
+* Don't mix source files from other projects in the same folder.
+* If keeping tests in the same folder as other source files, give them a distinct name so they can easily be excluded.
+* Avoid large build artifacts and dependency folders like `node_modules` in source directories
+
+Here is a reasonable `tsconfig.json` that demonstrates this in action.
+
+```json5
+{
+    "compilerOptions": {
+        // ...
+    },
+    "include": ["src"],
+    "exclude": ["**/node_modules", "**/.*/"],
+}
+```
+
+## Incremental Project Emit
+
+The `--incremental` flag allows TypeScript to save state from the last compilation to a `.tsbuildinfo` file.
+This file is used to figure out the smallest set of files that might to be re-checked/re-emitted since it last ran, much like how TypeScript's `--watch` mode works.
+
+Incremental compiles are enabled by default when using the `composite` flag for project references, but can bring the same speed-ups for any project that opts in.
```

Comment:
One thing I realized subsequent to writing this post:

 https://blog.johnnyreilly.com/2019/09/start-me-up-ts-loader-meet-tsbuildinfo.html

Is that `.tsbuildinfo` is only emitted in a `projectReferences` setup. That was surprising to me and to a bunch of others that commented on various ts-loader issues after the fact. 

I don't know if it will ever be available standalone? Either way, good that the knowledge is captured here 😀

Reply:
From looking at the commits it sounds like as of recently it's emitted, but only in `--watch` mode. (but maybe @sheetalkamat and @andrewbranch can clarify).

Reply:
I think you can use it with non-composite projects by passing `--incremental`.

Reply:
Sorry, I don't understand the question.  I'm pretty sure `--watch` works on non-composite projects, so I'm not sure what "standalone" means.  Maybe there's no API exposing it?